rfd/0205-improved-onprem-joining.md

```diff
@@ -0,0 +1,418 @@
+  makes it great for use in documentation - `tctl bots add` even gives you the
+  command to run to start a bot immediately!
+
+- Running very few, very reliable, long-lived systems. If you can reasonably
+  expect your system to never go down for more than 24 hours, bots can happily
+  run for months.
+
+  (Ironically, Kubernetes is a great environment in which to run `token`-joined
+  bots since it'll rapidly reschedule any bot deployments that fail... but we
+  have a dedicated `kubernetes` delegated join method.)
+
+End users willing to create their own automation around token issuance could
+work around some of these limitations, but this creates an unnecessary barrier
+to entry for use of Machine ID on-prem.
+
+## Details
+
+### Context: Priorities and Security Invariants
+
+It's become clear through conversations with customers and previous attempts at
+solving this problem that some of these pain points are contradictory. As a case
+study, an ideal UX from an end user's perspective might be to create one token,
+join all their bots with it, and leave them running indefinitely. This would
+create several issues:
+
+- The initial joining secret would have an ambiguous lifetime. At what point
+  does this multi-use token expire?
+
+- How many bots will join?
+
+- How can we tell joined bots apart? Can we trust a bot identity if it can be
+  thrown away and regenerated?
+
+- When a bot needs to rejoin, does it use the same token? Can that token *ever*
+  expire?
+
+With this in mind, we need to strike some balance between effective UX and a
+system we can trust to not allow unauthorized or unintended access. To that end,
+we'll focus on these explicit compromises we believe improve today's UX while
+making minimal security concessions:
+
+- **There must be a 1:1 relationship between a `ProvisionToken` and a bot
+  instance.** Allowing many joins on one token creates unnecessary backend
+  contention and - depending on implementation - creates severe traceability
+  problems.
+
+  However, we can greatly improve today's automation story. Secret fulfillment
+  can take place server side to reduce the number of resources to generate in
+  Terraform, and `ProvisionToken` resources can be made reusable to fully enable
+  IaC workflows.
+
+- **Bot credentials can be long-lived, but their *trust* must be controlled and
+  renewed.** We can't allow any secret values to either remain both valid and
+  useful for a long time, or allow currently valid identities to generate
+  credentials that can be used for future, unchecked extension of access.
+
+  However, we can create a new state where an identity is technically valid
+  indefinitely, but useless unless explicitly allowed by the server. Existing
+  controls like the generation counter can still effectively prevent unintended
+  reuse of the bot identity.
+
+### Challenge-Response Joining
+
+TODO: Consider alternative join method names?
```

Comment:
`bound-keypair`?

Reply:
Mentioning a key or keypair feels more intuitive to me because it's focussed on what the user has and sees rather than the technical implementation (the same is true of `token` and `tpm`), and will feel familiar if you've used SSH keys for authenticating with GitHub, etc.